build/Microsoft.DotNet.Cli.Test.targets

```diff
@@ -0,0 +1,157 @@
+    <ItemGroup>
+      <RestoreDesktopTestAssetProjectsInputs Include="$(RepoRoot)/TestAssets/TestProjects/**/project.json" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="BuildTestAssetProjects" >
+    <ItemGroup>
+      <NoAutoBuildTestAssets Include="$(RepoRoot)/TestAssets/TestProjects/**/.noautobuild" />
+      <NoAutoBuildTestAssetProjects Include="%(NoAutoBuildTestAssets.RelativeDir)/project.json" />
+      <BuildableTestAssetProjects Include="$(RepoRoot)/TestAssets/TestProjects/**/project.json" Exclude="@(NoAutoBuildTestAssetProjects)" />
+    </ItemGroup>
+
+    <Message Text="Skipping projects with .noautobuild files:" />
+    <Message Text="%(NoAutoBuildTestAssetProjects.FullPath)" />
+
+    <Exec Command="$(DotnetUnderTest) build %(BuildableTestAssetProjects.FullPath) --framework netcoreapp1.0" WorkingDirectory="%(BuildableTestAssetProjects.RelativeDir)" />
+  </Target>
+
+  <Target Name="BuildDesktopTestAssetProjects" Condition=" '$(IsDesktopAvailable)' == 'True' " >
+    <ItemGroup>
+      <NoAutoBuildDesktopTestAssets Include="$(RepoRoot)/TestAssets/TestProjects/**/.noautobuild" />
+      <NoAutoBuildDesktopTestAssetProjects Include="%(NoAutoBuildDesktopTestAssets.RelativeDir)/project.json" />
+      <BuildableDesktopTestAssetProjects Include="$(RepoRoot)/TestAssets/DesktopTestProjects/**/project.json" Exclude="@(NoAutoBuildTestAssetProjects)" />
+    </ItemGroup>
+
+    <Message Text="Skipping projects with .noautobuild files:" />
+    <Message Text="%(NoAutoBuildTestAssetProjects.FullPath)" />
+
+    <Exec Command="$(DotnetUnderTest) build %(BuildableTestAssetProjects.FullPath) --framework net451" WorkingDirectory="%(BuildableTestAssetProjects.RelativeDir)" />
+  </Target>
+
+  <Target Name="RestoreTests"
+          DependsOnTargets="SetupRestoreTestsInputs;"
+          Inputs="@(RestoreTestsInputs)"
+          Outputs="@(RestoreTestsInputs->'%(RelativeDir)/project.lock.json')">
+    <Exec Command="$(DotnetUnderTest) restore $(TestDirectory) --fallbacksource $(TestPackagesDir)" WorkingDirectory="$(TestDirectory)" />
+  </Target>
+
+  <Target Name="SetupRestoreTestsInputs"
+          DependsOnTargets="PrepareTests;" >
+    <ItemGroup>
+      <RestoreTestsInputs Include="$(TestDirectory)/**/project.json" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="BuildTests" 
+          DependsOnTargets="PrepareTests;">
+    <Exec Command="$(DotnetUnderTest) build --configuration $(Configuration)" WorkingDirectory="$(TestDirectory)/%(ProjectsToTest.Identity)/" />
+  </Target>
+
+  <Target Name="RunTests" 
+          DependsOnTargets="PrepareTests;">
+    <PropertyGroup>
+      <PathSeparator>:</PathSeparator>
+      <PathSeparator Condition=" '$(OSName)' == 'win' ">;</PathSeparator>
+      <ExecPath>$(DotnetStage2)$(PathSeparator)$(PATH)</ExecPath>
+      <TestArtifactsEnv>$(TestArtifactsDir)</TestArtifactsEnv>
+      <EnvPrefix>PATH=$(ExecPath) TEST_ARTIFACTS=$(TestArtifactsEnv)</EnvPrefix>
```

Comment:
If you use a ToolTask (i.e. derive a "DotNetTest" class from "DotNetTool"), you can use the EnvironmentVariables property:

[https://msdn.microsoft.com/en-us/library/microsoft.build.utilities.tooltask.environmentvariables.aspx](https://msdn.microsoft.com/en-us/library/microsoft.build.utilities.tooltask.environmentvariables.aspx)

(or since Exec derives from ToolTask, you should be able to use that property on Exec, but I don't see it documented on msdn)